src/Toml/BiMap.hs

```diff
@@ -51,245 +56,314 @@ module Toml.BiMap
-_Read :: (Show a, Read a) => BiMap a AnyValue
+_Read :: (Show a, Read a) => TomlBiMap a AnyValue
 _Read = _ReadString >>> _String
 
 -- | Helper bimap for 'Natural' and 'Integer'.
-_NaturalInteger :: BiMap Natural Integer
-_NaturalInteger = BiMap (Just . toInteger) maybeInteger
+_NaturalInteger :: TomlBiMap Natural Integer
+_NaturalInteger = BiMap (Right . toInteger) eitherInteger
   where
-    maybeInteger :: Integer -> Maybe Natural
-    maybeInteger n
-      | n < 0     = Nothing
-      | otherwise = Just (fromIntegral n)
+    eitherInteger :: Integer -> Either TomlBiMapError Natural
+    eitherInteger n
+      | n < 0     = Left $ ArbitraryError "Value is below zero."
+      | otherwise = Right (fromIntegral n)
 
 -- | 'Natural' bimap for 'AnyValue'. Usually used with 'natural' combinator.
-_Natural :: BiMap Natural AnyValue
+_Natural :: TomlBiMap Natural AnyValue
 _Natural = _NaturalInteger >>> _Integer
 
 -- | Helper bimap for 'Integer' and integral, bounded values.
-_BoundedInteger :: (Integral a, Bounded a) => BiMap a Integer
-_BoundedInteger = BiMap (Just . toInteger) maybeBounded
+_BoundedInteger :: (Integral a, Bounded a, Show a) => TomlBiMap a Integer
+_BoundedInteger = BiMap (Right . toInteger) eitherBounded
   where
-    maybeBounded :: forall a. (Integral a, Bounded a) => Integer -> Maybe a
-    maybeBounded n
-      | n < toInteger (minBound :: a) = Nothing
-      | n > toInteger (maxBound :: a) = Nothing
-      | otherwise                     = Just (fromIntegral n)
+    eitherBounded :: forall a. (Integral a, Bounded a, Show a) => Integer -> Either TomlBiMapError a
+    eitherBounded n
+      | n < toInteger (minBound :: a) =
+            Left $ ArbitraryError $ "Value " <> tShow n <> " is less minBound" <> tShow (minBound :: a)
+      | n > toInteger (maxBound :: a) =
+            Left $ ArbitraryError $ "Value " <> tShow n <> " is more maxBound" <> tShow (maxBound :: a)
+      | otherwise                     = Right (fromIntegral n)
 
 -- | 'Word' bimap for 'AnyValue'. Usually used with 'word' combinator.
-_Word :: BiMap Word AnyValue
+_Word :: TomlBiMap Word AnyValue
 _Word = _BoundedInteger >>> _Integer
 
 -- | 'Int' bimap for 'AnyValue'. Usually used with 'int' combinator.
-_Int :: BiMap Int AnyValue
+_Int :: TomlBiMap Int AnyValue
 _Int = _BoundedInteger >>> _Integer
 
 -- | 'Float' bimap for 'AnyValue'. Usually used with 'float' combinator.
-_Float :: BiMap Float AnyValue
+_Float :: TomlBiMap Float AnyValue
 _Float = iso realToFrac realToFrac >>> _Double
 
 -- | Helper bimap for 'Text' and strict 'ByteString'
-_ByteStringText :: BiMap ByteString Text
-_ByteStringText = prism T.encodeUtf8 maybeText
-  where
-    maybeText :: ByteString -> Maybe Text
-    maybeText = either (const Nothing) Just . T.decodeUtf8'
+_ByteStringText :: TomlBiMap ByteString Text
+_ByteStringText = prism T.encodeUtf8
+    (\value -> either (\_ -> leftWrongValue . Text $ tShow value) Right $ T.decodeUtf8' value)
```

Comment:
It's a good idea to do here the same with `ArbitraryError` as you did in `_LByteStringText`